src/mrfgen/mrfgen.py

```diff
@@ -589,15 +707,25 @@ def run_mrf_insert(mrf, tiles, insert_method, resize_resampling, target_x, targe
                                                               str((Decimal(t_xmax)-Decimal(t_xmin))/Decimal(target_x)),
                                                               str((Decimal(t_ymin)-Decimal(t_ymax))/Decimal(target_y)),
                                                               working_dir)
-
-            errors += run_mrf_insert(mrf, [left_half, right_half], insert_method, resize_resampling, target_x, target_y,
+            if should_lock:
+                lock.up_read()
+            errors += run_mrf_insert([left_half, right_half], mrf, insert_method, resize_resampling, target_x, target_y,
                                      mrf_blocksize, target_extents, target_epsg, nodata, True, working_dir)
             continue
 
         if merge: # merge tile with existing imagery if true
+            if should_lock:
+                lock.up_read()
+                lock.down_write()
+
             log_info_mssg("Image extents " + str([s_xmin, s_ymax, s_xmax, s_ymin]))
             tile = gdalmerge(mrf, tile, [s_xmin, s_ymax, s_xmax, s_ymin], target_x, target_y, mrf_blocksize,
```

Comment:
Any chance an exception would be thrown here and then cause an issue with locks in the wrong position?

Reply:
@jtroberts could correct me on this, but in (I think) all cases, errors are either caught and logged, or the entire program exits. If an error occurs that is handled, the `errors` variables is incremented and the locks are handled as usual. Otherwise sys.exit() is called. The lock is reset on each individual run.

Reply:
That sounds right to me.